File: 2025-2/sesion-de-clase-02/.NET/Ejercicio01/Persona.cs
using System;
using System.Runtime.InteropServices;

namespace Ejercicio01
{
    public abstract class Persona {
        public Persona() {
            Console.WriteLine("Persona: Constructor");
        }

        protected Persona(PersonaBuilder builder) {
            Nombre = builder.Nombre;
            Edad = builder.Edad;
        }

        public string Nombre { get; set; }
        public int Edad { get; set; }

        //private string nombre;

        //public string Nombre { 
        //    get { 
        //        return nombre;
        //    }
        //    set { 
        //        nombre = value;
        //    }
        //}

        public override string ToString() {
            return string.Format("Nombre: {0}, Edad: {1}", this.Nombre, this.Edad);
        }

        //public virtual void Imprimir() {
        //    Console.WriteLine("Nombre: {0}, Edad: {1}",
        //        this.Nombre, this.Edad);
        //}

        public abstract void Imprimir();

        ~Persona() {
            Console.WriteLine("Se destruye el objeto.");
        }

        public abstract class PersonaBuilder {
            public string Nombre { get; protected set; }
            public int Edad { get; protected set; }

            public virtual PersonaBuilder ConNombre(string nombre) {
                this.Nombre = nombre;
                return this;
            }

            public virtual PersonaBuilder ConEdad(int edad) { 
                this.Edad = edad;
                return this;
            }

            public abstract Persona Build();
        }
    }
}

File: 2025-2/sesion-de-clase-02/Java/ejercicio01/src/main/java/pe/edu/pucp/inf30/ejercicio01/Persona.java
package pe.edu.pucp.inf30.ejercicio01;

/**
 *
 * @author eric
 */
public abstract class Persona {
    private String nombre;
    private int edad;
    
    public Persona() {
        System.out.println("Persona: Constructor");
    }
    
    public String getNombre() {
        return this.nombre;
    }
    
    public int getEdad() {
        return this.edad;
    }

    public void setNombre(String nombre) {
        this.nombre = nombre;
    }
    
    public void setEdad(int edad) {
        this.edad = edad;
    }
    
    Persona(PersonaBuilder builder) {
        this.nombre = builder.getNombre();
        this.edad = builder.getEdad();
    }
    
    @Override
    public String toString() {
        return String.format("Nombre: %s, Edad: %d", this.nombre, this.edad);
    }
    
    public abstract void imprimir();
    
    public abstract static class PersonaBuilder {
        protected String nombre;
        protected int edad;
        
        public String getNombre() {
            return this.nombre;
        }
        
        public int getEdad() {
            return this.edad;
        }
        
        public PersonaBuilder nombre(String nombre) {
            this.nombre = nombre;
            return this;
        }
        
        public PersonaBuilder edad(int edad) {
            this.edad = edad;
            return this;
        }
        
        public abstract Persona build();
    }
}
